src/sys.rs

```diff
@@ -0,0 +1,374 @@
+    pub error: bool,
+}
+
+impl Readiness {
+    /// Shorthand for empty readiness
+    pub const EMPTY: Readiness = Readiness {
+        readable: false,
+        writable: false,
+        error: false,
+    };
+}
+
+#[derive(Debug)]
+pub(crate) struct PollEvent {
+    pub(crate) readiness: Readiness,
+    pub(crate) token: Token,
+}
+
+/// Factory for creating tokens in your registrations
+///
+/// When composing event sources, each sub-source needs to
+/// have its own token to identify itself. This factory is
+/// provided to produce such unique tokens.
+
+#[derive(Debug)]
+pub struct TokenFactory {
+    /// The key of the source this factory is associated with.
+    key: usize,
+
+    /// The next sub-id to use.
+    sub_id: u32,
+}
+
+impl TokenFactory {
+    pub(crate) fn new(key: usize) -> TokenFactory {
+        TokenFactory { key, sub_id: 0 }
+    }
+
+    /// Produce a new unique token
+    pub fn token(&mut self) -> Token {
+        // Ensure we don't overflow the sub-id.
+        if self.sub_id >= MAX_SUBSOURCES_TOTAL as _ {
+            panic!("Too many sub-sources for this source");
+        }
+
+        // Compose the key and the sub-key together.
+        let mut key = self.key;
+        key |= (self.sub_id as usize) << MAX_SOURCES;
+
+        let token = Token { key };
+
+        self.sub_id += 1;
+
+        token
+    }
+}
+
+/// A token (for implementation of the [`EventSource`](crate::EventSource) trait)
+///
+/// This token is produced by the [`TokenFactory`] and is used when calling the
+/// [`EventSource`](crate::EventSource) implementations to process event, in order
+/// to identify which sub-source produced them.
+///
+/// You should forward it to the [`Poll`] when registering your file descriptors.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Token {
+    pub(crate) key: usize,
+}
+
+/// The polling system
+///
+/// This type represents the polling system of calloop, on which you
+/// can register your file descriptors. This interface is only accessible in
+/// implementations of the [`EventSource`](crate::EventSource) trait.
+///
+/// You only need to interact with this type if you are implementing your
+/// own event sources, while implementing the [`EventSource`](crate::EventSource) trait.
+/// And even in this case, you can often just use the [`Generic`](crate::generic::Generic) event
+/// source and delegate the implementations to it.
+pub struct Poll {
+    /// The handle to wepoll/epoll/kqueue/... used to poll for events.
+    poller: Arc<Poller>,
+
+    /// The buffer of events returned by the poller.
+    events: RefCell<Vec<Event>>,
+
+    pub(crate) timers: Rc<RefCell<TimerWheel>>,
+}
+
+impl std::fmt::Debug for Poll {
+    #[cfg_attr(coverage, no_coverage)]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("Poll { ... }")
+    }
+}
+
+impl Poll {
+    pub(crate) fn new(_high_precision: bool) -> crate::Result<Poll> {
```

Comment:
`polling` currently just does "high precision" polling all the time when it can. If that isn't desired I can add a method in `polling` to allow creating a `Poller` without a high-precision timer.

Reply:
If high precision polling is always active, then this argument is now meaningless and should be removed.